src/main/java/org/broadinstitute/hellbender/engine/spark/ShuffleJoinReadsWithRefBases.java

```diff
@@ -80,4 +81,34 @@
             return out;
         });
     }
-}
\ No newline at end of file
+
+    public static <T> JavaPairRDD<GATKRead, Tuple2<T, ReferenceBases>> addBases(final ReferenceDataflowSource referenceDataflowSource,
+                                                                                final JavaPairRDD<GATKRead, T> keyedByRead) {
+        SerializableFunction<GATKRead, SimpleInterval> windowFunction = referenceDataflowSource.getReferenceWindowFunction();
+
+        JavaPairRDD<ReferenceShard, Tuple2<GATKRead, T>> shardRead = keyedByRead.mapToPair(pair -> {
+            ReferenceShard shard = ReferenceShard.getShardNumberFromInterval(windowFunction.apply(pair._1()));
+            return new Tuple2<>(shard, pair);
+        });
+
+        JavaPairRDD<ReferenceShard, Iterable<Tuple2<GATKRead, T>>> shardiRead = shardRead.groupByKey();
+
+        return shardiRead.flatMapToPair(in -> {
+            List<Tuple2<GATKRead, Tuple2<T, ReferenceBases>>> out = Lists.newArrayList();
+            Iterable<Tuple2<GATKRead, T>> iReads = in._2();
+
+            // Apply the reference window function to each read to produce a set of intervals representing
+            // the desired reference bases for each read.
+            final List<SimpleInterval> readWindows = StreamSupport.stream(iReads.spliterator(), false).map(pair -> windowFunction.apply(pair._1())).collect(Collectors.toList());
+
+            SimpleInterval interval = SimpleInterval.getSpanningInterval(readWindows);
+            // TODO: don't we need to support GCS PipelineOptions?
+            ReferenceBases bases = referenceDataflowSource.getReferenceBases(null, interval);
```

Comment:
For Spark, we bake the apiKey from PipelineOptions into the referenceDataflowSource (RefAPISource).